main/src/ca/uwaterloo/flix/language/phase/Resolver.scala

```diff
@@ -1356,12 +1356,19 @@ object Resolver {
             case Result.Ok(st0) =>
               flatMapN(getStructIfAccessible(st0, ns0, loc)) {
                 case st =>
-                  val providedFieldNames = fields.map { case (k, _) => k }.toSet
-                  val expectedFieldNames = st.fields.map(_.name).toSet
+                  val providedFieldNames = fields.map { case (k, _) => k }
+                  val expectedFieldNames = st.fields.map(_.name)
                   val extraFields = providedFieldNames.diff(expectedFieldNames)
                   val missingFields = expectedFieldNames.diff(providedFieldNames)
 
-                  val errors = extraFields.map(ResolutionError.ExtraStructField(st0.sym, _, loc)) ++ missingFields.map(ResolutionError.MissingStructField(st0.sym, _, loc))
+                  val errors = if (!extraFields.isEmpty || !missingFields.isEmpty) {
```

Comment:
I understand the logic here and I agree its correct. But it is still a bit hard to follow.

I wonder if we could spell it out:

if (expected == actual) {
  // all god
 } else {
  // Three things can be wrong:
  // Case 1: Missing Fields
  if (missingFields.nonEmpty) ...
  // Case 2: Extra fields
  if (extrafields.nonEmpty) { ... }

}

Reply:
OK; I am not sure. Maybe you can have a go at it ?

Reply:
I suppose a tricky thing is that you want to report both extra and missing fields at the same time.